ext/CEGUI/IrrlichtBaWTexture.cpp

```diff
@@ -220,72 +152,117 @@ void IrrlichtBaWTexture::loadFromMemory(const void* buffer, const CEGUI::Sizef&
-        driver->blitRenderTargets(temporaryInFrameBuffer, frameBuffer, false, false, coverArea, coverArea);
-        temporaryInFrameBuffer->drop();
+        auto requirements = getMemoryRequirements(area.getWidth() * area.getHeight() * asset::getTexelOrBlockBytesize(viewParams.format));
+
+        auto gpuBuffer = driver->createGPUBufferOnDedMem(requirements);
+        driver->updateBufferRangeViaStagingBuffer(gpuBuffer.get(), 0, requirements.vulkanReqs.size, sourceData);
+
+        auto regions = core::make_refctd_dynamic_array<core::smart_refctd_dynamic_array<asset::ICPUImage::SBufferCopy>>(1u);
+        asset::ICPUImage::SBufferCopy& region = regions->front();
+        region.imageSubresource.mipLevel = 0u;
+        region.imageSubresource.baseArrayLayer = 0u;
+        region.imageSubresource.layerCount = 1u;
+        region.bufferOffset = 0u;
+        region.bufferRowLength = 0;
+        region.bufferImageHeight = 0u;
+        region.imageOffset = { area.left(), area.top(), 0u };
+        region.imageExtent = extent;
+
+        driver->copyBufferToImage(gpuBuffer.get(), gpuImageViewTexture->getCreationParameters().image.get(), regions->size(), regions->begin());
     }
     else
-        assert(status);
+        assert(status, _IRR_TEXTURE_MUST_EXIST_.data());
 }
 
 void IrrlichtBaWTexture::blitToMemory(void* targetData)
 {
-    bool status = frameBuffer;
+    bool status = gpuImageViewTexture.get();
 
     if (status)
     {
-        auto gpuTexture = frameBuffer->getAttachment(textureColorAttachment);
-        auto& viewParams = gpuTexture->getCreationParameters();
+        auto& viewParams = gpuImageViewTexture->getCreationParameters();
         auto& extent = viewParams.image->getCreationParameters().extent;
-
-        auto temporaryOutFrameBuffer = createFrameBuffer(targetData, viewParams.format, extent.width, extent.height);
-        driver->blitRenderTargets(frameBuffer, temporaryOutFrameBuffer, false, false);
- 
-        auto bufferByteSize = viewParams.image->getImageDataSizeInBytes();
-        auto destinationBoundMemory = const_cast<video::IDriverMemoryAllocation*>(viewParams.image->getBoundMemory()); // not sure
-        destinationBoundMemory->mapMemoryRange(video::IDriverMemoryAllocation::EMCAF_READ, { 0u, bufferByteSize });
-        core::smart_refctd_ptr<asset::ICPUBuffer> pixelBuffer = core::make_smart_refctd_ptr<asset::CCustomAllocatorCPUBuffer<core::null_allocator<uint8_t>>>(bufferByteSize, destinationBoundMemory->getMappedPointer(), core::adopt_memory);
-
-        memcpy(targetData, pixelBuffer->getPointer(), pixelBuffer->getSize());
-
-        temporaryOutFrameBuffer->drop();
+        const auto bufferByteSize = asset::getTexelOrBlockBytesize(viewParams.format) * extent.width * extent.height;
+
+        video::IDriverMemoryBacked::SDriverMemoryRequirements requirements = getMemoryRequirements(bufferByteSize);
+        auto gpuBuffer = driver->createGPUBufferOnDedMem(requirements);
```

Comment:
don't create a new buffer, we have the downstreaming buffer, same thing we use in the screenshot extension!